pep-0688.rst

```diff
@@ -0,0 +1,258 @@
+no backwards compatibility concerns.
+
+However, the recommendation to remove the special behavior for
+``bytes`` in type checkers does have backwards compatibility
+impact on users. An `experiment <https://github.com/python/mypy/pull/12661>`__
+with mypy shows that several major open source projects type
+checked with mypy will see new errors if the ``bytes`` promotion
+is removed. However, many of these errors can be fixed by improving
+the stubs in typeshed, as already done for the
+`builtins <https://github.com/python/typeshed/pull/7631>`__,
+`binascii <https://github.com/python/typeshed/pull/7677>`__,
+`pickle <https://github.com/python/typeshed/pull/7678>`__ modules.
+Overall, the change improves overall type safety,
+so we believe the migration cost is worth it.
+
+
+Security Implications
+=====================
+
+None.
+
+
+How to Teach This
+=================
+
+We will add notes pointing to ``types.Buffer`` to appropriate places in the
+documentation, such as `typing.readthedocs.io <https://typing.readthedocs.io/en/latest/>`__
+and the `mypy cheat sheet <https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html>`__.
+Type checkers may provide additional pointers in their error messages. For example,
+when they encounter a place where a buffer object is passed to a function that
+is annotated to only accept ``bytes``, the error message could include a note suggesting
+to use ``types.Buffer`` instead.
+
+
+Reference Implementation
+========================
+
+An implementation of ``types.Buffer`` is
+`available <https://github.com/python/cpython/compare/main...JelleZijlstra:typesbuffer?expand=1>`__
+in the author's fork.
+
+
+Rejected Ideas
+==============
+
+Buffer ABC
+----------
+
+An `earlier proposal <https://github.com/python/cpython/issues/71688>`__ suggested
+adding a ``collections.abc.Buffer`` ABC to represent buffer objects. This idea
+stalled because an ABC with no methods does not fit well into the ``collections.abc``
+module. Furthermore, it required manual registration of buffer classes, including
+those in the standard library. This PEP's approach of using the ``__instancecheck__``
+hook is more natural and does not require explicit registration.
+Nevertheless, the ABC proposal has the advantage that it does not require C changes,
+and we are proposing to adopt a version of it in the third-party ``typing_extensions``
+package for the benefit of users of older versions of Python.
+
+
+Open Issues
+===========
+
+Read-only and writable buffers
+------------------------------
+
+To avoid making changes to the buffer protocol itself, this PEP currently
+does not provide a way to distinguish between read-only and writable buffers.
+That's unfortunate, because some APIs require a writable buffer, and one of
+the most common buffer types (``bytes``) is always read-only.
+Should we add a new mechanism in C to declare that a type implementing the
+buffer protocol is always read-only?
```

Comment:
Sidenote: At least at the level of static type checking, wouldn't the opposite make more sense, i.e. declaring that a type is writable, since a writable buffer is a subtype of a read-only buffer? This also follows the general principle that for the base collections, etc. e.g. `Set`, `Mapping`, `Sequence`, etc, the base type is read-only, with mutable subtypes. But I'm not a typing expert like you, and there may be other C-implementation reasons this framing makes more sense.

Reply:
That's a good point actually. At the Python level, I'd want a WritableBuffer subclass of Buffer that bytearray implements but bytes doesn't. I'll reword the question to point in that direction.